File: programmers/level1/키패드_누르기/solution.py
import math

# 유클리드 거리 구하기 (피타고라스 공식이랑 비슷한 것 같다)
def euclidean_distance(coord_one, coord_two):
    return math.sqrt(abs(coord_one[0] - coord_two[0]) + abs(coord_one[1] - coord_two[1]))


# 맨해튼 거리 (이 문제에선 이 공식이 맞는 것 같다)
def manhattan_distance(coord_one, coord_two):
    return abs(coord_one[0] - coord_two[1]) + abs(coord_one[1] + coord_two[1])


def solution(numbers, hand):
    answer = ""

    right_thumb_loc = "#"
    left_thumb_loc = "*"

    key_pad = {
        "1": [0, 0],
        "2": [0, 1],
        "3": [0, 2],
        "4": [1, 0],
        "5": [1, 1],
        "6": [1, 2],
        "7": [2, 0],
        "8": [2, 1],
        "9": [2, 2],
        "*": [3, 0],
        "0": [3, 1],
        "#": [3, 2],
    }

    thumbs = []
    for number in numbers:
        if number in [1, 4, 7]:
            left_thumb_loc = str(number)
            thumbs.append("L")
        elif number in [3, 6, 9]:
            right_thumb_loc = str(number)
            thumbs.append("R")
        else:
            right_dist = manhattan_distance(key_pad.get(right_thumb_loc), key_pad.get(str(number)))
            left_dist = manhattan_distance(key_pad.get(left_thumb_loc), key_pad.get(str(number)))

            if right_dist < left_dist:
                thumbs.append("R")
                right_thumb_loc = str(number)
            elif left_dist < right_dist:
                thumbs.append("L")
                left_thumb_loc = str(number)
            else:
                if hand == "right":
                    thumbs.append("R")
                    right_thumb_loc = str(number)
                else:
                    thumbs.append("L")
                    left_thumb_loc = str(number)

    answer = "".join(thumbs)

    return answer


if __name__ == "__main__":
    print(solution([1, 3, 4, 5, 8, 2, 1, 4, 5, 9, 5], "right"))

File: programmers/level1/키패드_누르기/Solution.java
import java.util.*;

public class Solution {
    // 유클리드 거리 구하기 (피타고라스 공식이랑 비슷한 것 같다)
    public double euclideanDistance(int[] coordOne, int[] coordTwo) {
        return Math.sqrt(Math.abs(coordOne[0] - coordTwo[0]) + Math.abs(coordOne[1] - coordTwo[1]));
    }

    // 맨해튼 거리 (이 문제에선 이 공식이 맞는 것 같다)
    public int manhattanDistance(int[] coordOne, int[] coordTwo) {
        return Math.abs(coordOne[0] - coordTwo[0]) + Math.abs(coordOne[1] - coordTwo[1]);
    }

    public String solution(int[] numbers, String hand) {
        StringBuilder rightThumbLoc = new StringBuilder("*");
        StringBuilder leftThumbLoc = new StringBuilder("#");

        Map<String, int[]> keyPad = new HashMap<>();

        keyPad.put("1", new int[] { 0, 0 });
        keyPad.put("2", new int[] { 0, 1 });
        keyPad.put("3", new int[] { 0, 2 });
        keyPad.put("4", new int[] { 1, 0 });
        keyPad.put("5", new int[] { 1, 1 });
        keyPad.put("6", new int[] { 1, 2 });
        keyPad.put("7", new int[] { 2, 0 });
        keyPad.put("8", new int[] { 2, 1 });
        keyPad.put("9", new int[] { 2, 2 });
        keyPad.put("*", new int[] { 3, 0 });
        keyPad.put("0", new int[] { 3, 1 });
        keyPad.put("#", new int[] { 3, 2 });

        List<String> thumbs = new ArrayList<>();

        for (int number : numbers) {
            if (number == 1 || number == 4 || number == 7) {
                leftThumbLoc.replace(0, leftThumbLoc.length(), Integer.toString(number));
                thumbs.add("L");
            } else if (number == 3 || number == 6 || number == 9) {
                rightThumbLoc.replace(0, rightThumbLoc.length(), Integer.toString(number));
                thumbs.add("R");
            } else {
                int rightDist = manhattanDistance(keyPad.get(rightThumbLoc.toString()),
                        keyPad.get(Integer.toString(number)));
                int leftDist = manhattanDistance(keyPad.get(leftThumbLoc.toString()),
                        keyPad.get(Integer.toString(number)));

                if (rightDist < leftDist) {
                    thumbs.add("R");
                    rightThumbLoc.replace(0, rightThumbLoc.length(), Integer.toString(number));
                } else if (leftDist < rightDist) {
                    thumbs.add("L");
                    leftThumbLoc.replace(0, leftThumbLoc.length(), Integer.toString(number));
                } else {
                    if (hand.equals("right")) {
                        thumbs.add("R");
                        rightThumbLoc.replace(0, rightThumbLoc.length(), Integer.toString(number));
                    } else {
                        thumbs.add("L");
                        leftThumbLoc.replace(0, leftThumbLoc.length(), Integer.toString(number));
                    }
                }
            }
        }

        String answer = String.join("", thumbs);

        return answer;
    }

    public static void main(String[] args) {
        System.out.println(new Solution().solution(new int[] { 1, 3, 4, 5, 8, 2, 1, 4, 5, 9, 5 }, "right"));
    }
}